### PR TITLE
[crypto] poseidon.js

### PR DESCRIPTION
**Pull request by mimoo**
This commit exports an OCaml poseidon hash implemention to javascript.
It uses the fp_3 parameters (see https://github.com/o1-labs/marlin/blob/master/oracle/src/poseidon.rs#L73).
In particular, the javascript implementation has two methods:

- `hashBytearray` to hash an opaque array of bytes. The type expected is a Uint8Array.
- `hashFieldElems` to hash an array of field elements. The field elements must be represented in little-endian hex encoded form.

Both methods return a digest in hexadecimal form of a little-endian encoded field element.

For more details see https://hackmd.io/ecBTibDpTzGdH9WM30gwQw

### CODE REVIEW

**Review by imeckler — COMMENTED** [review]
Looks great! Will approve after test vectors are in :slightly_smiling_face:

**Comment by mimoo** [comment]
test vectors pass! Someone needs to approve this on Snarky before this PR can be merged: https://github.com/o1-labs/snarky/pull/580

File: frontend/client_sdk/yarn.lock [hunk 1]
```diff
@@ -12,12 +12,12 @@ bs-platform@^7.0.2-dev.2:
...
+  integrity sha512-EKZzbS4HBuw80zDm5ZJxDdfoGw4nYClTgkv4vp8/db1MXmvCdJSA2NuvxNBYF8i5VqYccdQWzNGClmhWpTFWFQ==
 
-typescript@^3.7.5:
-  version "3.7.5"
-  resolved "https://registry.yarnpkg.com/typescript/-/typescript-3.7.5.tgz#0692e21f65fd4108b9330238aac11dd2e177a1ae"
-  integrity sha512-/P5lkRXkWHNAbcJIiHPfRoKqyd7bsyCma1hZNUGfn20qm64T6ZBlrzprymeu918H+mB/0rIg2gGK/BXkhhYgBw==
+typescript@^4.2.4:
+  version "4.2.4"
+  resolved "https://registry.yarnpkg.com/typescript/-/typescript-4.2.4.tgz#8610b59747de028fda898a8aef0e103f156d0961"
+  integrity sha512-V+evlYHZnQkaz8TRBuxTA92yZBPotr5H+WhQ7bD3hZUndx5tGOa1fuCgeSjxAzM1RiN5IzvadIXTVefuuwZCRg==
```
Comment (mimoo) [inline frontend/client_sdk/yarn.lock:23]: looks like I somehow automagically updated these, let's see if CI pass I guess?

File: src/app/client_sdk/dune [hunk 2]
```diff
@@ -1,10 +1,25 @@
...
+  data_hash_lib_nonconsensus
+  random_oracle_nonconsensus
+  signature_lib_nonconsensus
+  zarith_stubs_js
+  integers
+  integers_stubs_js
+  js_of_ocaml
+  digestif.ocaml
+  sponge
+  hex)
```
Comment (mimoo) [inline src/app/client_sdk/dune:20]: added sponge and hex

**Comment by mimoo** [comment]
!ci-build-me

**Comment by mimoo** [comment]
!ci-build-me

**Pull request closed without merging**